### ISSUE TITLE
Feature request: Javascript

### ISSUE THREAD

**Issue by flackr** [ArnoldSmith86/virtualtabletop#1609]
# Overview

I would love to be able to write game scripts in Javascript. It can be much shorter and easier to write, and significantly faster to execute. I'm opening this issue to gather ideas and discussion about how to make this happen. I think at a high level the goals are:

1. Access to the game state should have roughly the same capabilities as the JSON scripting. This means that supporting Javascript wouldn't add any API surface area to what developers can access. In particular, we wouldn't give direct access to the widget objects but rather have the same get and set property apis that go through the same mechanism that GET and SET do.
2. Ensure the scripting support is isolated and doesn't interfere with the core virtualtabletop script. In particular, the game author script should run in a [sandboxed iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox) and use message passing to interact with the main game.

# Design

## Where does the script live?

There's a few options here, and we may want to support multiple:

1. Architecturally, it would be easiest to load a separate file. This is nice because that file could be plain JS and wouldn't need to worry about being embedded in a JSON structure. The question would be where in the game that data lives. Files are somewhat inconvenient right now as any time you change them the asset changes, though perhaps asset aliases #1209 would make this easy.

   The main game `_meta` widget could have an attribute listing scripts. This would also help with detecting and integrating those scripts into the editor.

2. Javascript could be added the same way as JSON script as strings in the JSON, possibly even a naming convention like `myRoutineJS` would signify that the routine code is javascript. The downsides are that it would still have to be parseable JSON so it could require lots of escaped syntax, it would be slow if each time we encounter these routines we have to re-send the code to the frame to be evaluated - though we could possibly reuse the function if the code doesn't change.

## How does the script get invoked from a game

The simplest starting approach would be for entry into javascript to always come from a JSON scripting method. This would be similar to the WebAssembly model and only require the implementation of a single function.

E.g.
```
  "clickRoutine": [{"func": "CALL", "type": "javascript", "routine": "jsRoutineName", ...}]
```

Optionally we could shorten this by allowing an object structure in place of JSON routines, e.g.:
```
  "clickRoutine": {"function": "jsRoutineName"}
```

Though it may still be useful to have the call syntax for use by other JSON scripts.

## How does the script interact with the game

We would create JS classes whose methods work by posting messages over to the main frame to perform the change. E.g.

```js
let listeners = [];
let sequenceId = 0;
window.addEventListener('message', (e) => {
  listeners[e.data.sequence].apply(null, e.data.value);
});

class Widget {
  get(property) {
    return Promise((res) => {
      listeners[sequenceId] = res;
      window.opener.postMessage({method: 'GET', widget: this.id, property, sequence: sequenceId++});
    });
  }
}
```

This way the game author script could do things like:
```js
function doSomeWork(widget) {
  let previous = await widget.get('foo');
  await widget.set('foo', 'bar');
}
```

We should consider whether we want to do something more efficient that would let us batch updates. E.g. if we sent the full game state to the script, we could gather deltas and post them back to the client in a single message rather than individual messages for every change.

**Comment by robartsd** [comment]
> We should consider whether we want to do something more efficient that would let us batch updates. E.g. if we sent the full game state to the script, we could gather deltas and post them back to the client in a single message rather than individual messages for every change.

Full game state is not sufficient to fully replicate Widget.get() behavior; engine logic has many paths to provide values that are not directly in the state; in some instances that could require the game JS script to re-implement significant engine code (which could break as the engine evolves - we currently attempt mitigate this with fileupdater for JSON routines; but would be unable to do so for JS).

**Comment by flackr** [comment]
Right, that's a good point. It's honestly not that inefficient to send lots of messages between windows so it shouldn't be a problem.

I agree, fixing calls in the JS code would be significantly more challenging. Hypothetically we could version the calls and do the fixup at runtime but its nowhere near as clean as updating when loading so I wouldn't seriously suggest doing this.
